File: src/python/bots/semantic_kernel_bot.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

import os
from botbuilder.core import ConversationState, TurnContext, UserState, CardFactory
from botbuilder.schema import ChannelAccount, Activity, ActivityTypes
from openai import AzureOpenAI
import semantic_kernel as sk
from semantic_kernel.connectors.ai.open_ai import (
    AzureAISearchDataSource,
    AzureChatCompletion,
    AzureChatPromptExecutionSettings,
    ExtraBody,
    ApiKeyAuthentication
)
from semantic_kernel.connectors.memory.azure_cognitive_search.azure_ai_search_settings import AzureAISearchSettings
from semantic_kernel.contents import ChatHistory
from semantic_kernel.functions import KernelArguments
from semantic_kernel.prompt_template import InputVariable, PromptTemplateConfig
    
from data_models import ConversationData, ConversationTurn
from .state_management_bot import StateManagementBot
from utils import get_citations_card, replace_citations


class SemanticKernelBot(StateManagementBot):

    def __init__(self, conversation_state: ConversationState, user_state: UserState, aoai_client: AzureOpenAI):
        super().__init__(conversation_state, user_state)
        self._aoai_client = aoai_client

    async def on_members_added_activity(
        self, members_added: list[ChannelAccount], turn_context: TurnContext
    ):
        for member in members_added:
            if member.id != turn_context.activity.recipient.id:
                await turn_context.send_activity("Hello and welcome to the Semantic Kernel Bot!")

    async def on_message_activity(self, turn_context: TurnContext):
        # Get the state properties from the turn context.
        conversation_data = await self.conversation_data_accessor.get(
            turn_context, ConversationData([ConversationTurn(role="system", content=os.getenv("LLM_INSTRUCTIONS"))])
        )

        conversation_data.add_turn("user", turn_context.activity.text)

        kernel = sk.Kernel()

        az_source = AzureAISearchDataSource(parameters={
            "endpoint": os.getenv("AZURE_SEARCH_API_ENDPOINT"),
            "authentication": ApiKeyAuthentication(key=os.getenv("AZURE_SEARCH_API_KEY")),
            "index_name": os.getenv("AZURE_SEARCH_INDEX"),
            "query_type": "simple",
        })
        extra = ExtraBody(data_sources=[az_source])
        req_settings = AzureChatPromptExecutionSettings(service_id="default", extra_body=extra)

        chat_service = AzureChatCompletion(
            service_id="chat-gpt",
            api_key=os.getenv("AZURE_OPENAI_API_KEY"),
            endpoint=os.getenv("AZURE_OPENAI_API_ENDPOINT"),
            deployment_name=os.getenv("AZURE_OPENAI_DEPLOYMENT_NAME"),
        )
        kernel.add_service(chat_service)

        prompt_template_config = PromptTemplateConfig(
            template="{{$chat_history}}",
            name="chat",
            template_format="semantic-kernel",
            input_variables=[
                InputVariable(name="chat_history", description="The history of the conversation", is_required=True)
            ],
            execution_settings={"default": req_settings}
        )

        history = ChatHistory()

        for message in conversation_data.history:
            if message.role == "user":
                history.add_user_message(message.content)
            else:
                history.add_assistant_message(message.content)

        chat_function = kernel.add_function(
            plugin_name="ChatBot", function_name="Chat", prompt_template_config=prompt_template_config
        )

        arguments = KernelArguments(settings=req_settings)

        arguments["chat_history"] = history
        answer = await kernel.invoke(
            function=chat_function,
            arguments=arguments,
        )

        response = str(answer)
        response = replace_citations(response)

        conversation_data.add_turn("assistant", response)
        await turn_context.send_activity(response)

        if 'citations' in answer.get_inner_content().choices[0].message.context and len(answer.get_inner_content().choices[0].message.context['citations']) > 0:
            citations = answer.get_inner_content().choices[0].message.context['citations']
            await turn_context.send_activity(get_citations_card(citations))

File: src/dotnet/Bots/SemanticKernelBot.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using Plugins;

namespace GenAIBot.Bots
{
    public class SemanticKernelBot<T> : StateManagementBot<T> where T : Dialog
    {
        private readonly string _instructions;
        private readonly string _welcomeMessage;
        private readonly string _endpoint;
        private readonly string _deployment;
        private readonly AzureSearchChatDataSource _chatDataSource;
        private readonly TokenCredential _credential;

        public SemanticKernelBot(IConfiguration config, ConversationState conversationState, UserState userState, AzureOpenAIClient aoaiClient, AzureSearchChatDataSource chatDataSource, T dialog)
            : base(config, conversationState, userState, dialog)
        {
            _instructions = config["LLM_INSTRUCTIONS"];
            _welcomeMessage = config.GetValue("LLM_WELCOME_MESSAGE", "Hello and welcome to the Semantic Kernel Bot Dotnet!");
            _endpoint = config["AZURE_OPENAI_API_ENDPOINT"];
            _deployment = config["AZURE_OPENAI_DEPLOYMENT_NAME"];
            _credential = new DefaultAzureCredential();
            _chatDataSource = chatDataSource;
        }

        // Modify onMembersAdded as needed
        protected override async Task OnMembersAddedAsync(IList<ChannelAccount> membersAdded, ITurnContext<IConversationUpdateActivity> turnContext, CancellationToken cancellationToken)
        {
            foreach (var member in membersAdded)
            {
                if (member.Id != turnContext.Activity.Recipient.Id)
                {
                    await turnContext.SendActivityAsync(MessageFactory.Text(_welcomeMessage), cancellationToken);
                }
            }
            // Log in at the start of the conversation
            await HandleLogin(turnContext, cancellationToken);
        }

        protected override async Task<bool> OnMessageActivityAsync(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
        {
            // Load conversation state
            var conversationStateAccessors = _conversationState.CreateProperty<ConversationData>(nameof(ConversationData));
            var conversationData = await conversationStateAccessors.GetAsync(
                turnContext, () => new ConversationData()
                {
                    History = new List<ConversationTurn>() {
                        new() { Role = "system", Message = _instructions }
                    }
                });

            // Enforce login
            var loggedIn = await HandleLogin(turnContext, cancellationToken);
            if (!loggedIn)
            {
                return false;
            }

            // Add user message to history
            conversationData.AddTurn("user", turnContext.Activity.Text);

            var kernel = Kernel.CreateBuilder()
                .AddAzureOpenAIChatCompletion(_deployment, _endpoint, _credential)
                .Build();

            // Add custom plugins
            kernel.Plugins.Add(kernel.CreatePluginFromObject(new WikipediaPlugin(conversationData, turnContext)));

            var promptExecutionSettings = new AzureOpenAIPromptExecutionSettings()
            {
                ToolCallBehavior = ToolCallBehavior.AutoInvokeKernelFunctions
            };

            var history = new ChatHistory();
            foreach (var turn in conversationData.History)
            {
                if (turn.Role == "user")
                {
                    history.AddUserMessage(turn.Message);
                }
                else
                {
                    history.AddAssistantMessage(turn.Message);
                }
            }

            var chatCompletionService = kernel.GetRequiredService<IChatCompletionService>();

            var answer = await chatCompletionService.GetChatMessageContentAsync(history, promptExecutionSettings, kernel);

            var response = answer.ToString();

            // Add assistant message to history
            conversationData.AddTurn("assistant", response);

            // Respond back to user
            await turnContext.SendActivityAsync(MessageFactory.Text(response), cancellationToken);

            // Return true if bot should run next handlers
            return true;
        }
    }
}